### PR TITLE
Master persist more states to etcd, schedule pending timeout after lo…

### PR DESCRIPTION
**Pull request by helinwang**
…ad pending state.

### CODE REVIEW

File: go/master/service.go [hunk 1]
```diff
@@ -77,11 +77,13 @@ type taskEntry struct {
...
-	Pending map[int]taskEntry // map from task ID to task entry
-	Done    []taskEntry
-	Failed  []taskEntry
+type masterState struct {
+	Todo     []taskEntry
+	Pending  map[int]taskEntry // map from task ID to task entry
+	Done     []taskEntry
+	Failed   []taskEntry
+	CurPass  int
+	JobTasks []taskEntry
```
Comment (gongweibao) [inline go/master/service.go]: 赞JobTasks

File: go/master/service.go [hunk 2]
```diff
@@ -350,10 +357,10 @@ func (s *Service) checkTimeoutFunc(taskID int, epoch int) func() {
...
 func (s *Service) logFields() log.Fields {
 	return log.Fields{
-		"todoLen":    len(s.taskQueues.Todo),
-		"pendingLen": len(s.taskQueues.Pending),
-		"doneLen":    len(s.taskQueues.Done),
-		"failedLen":  len(s.taskQueues.Failed),
+		"todoLen":    len(s.state.Todo),
+		"pendingLen": len(s.state.Pending),
+		"doneLen":    len(s.state.Done),
+		"failedLen":  len(s.state.Failed),
```
Comment (typhoonzero) [inline go/master/service.go]: Also add CurrPass to logFields?

File: go/master/service.go (hunk 1 above)
Comment (typhoonzero) [inline go/master/service.go]: JobTasks stays the same across one job, maybe no need to save it every time? Anyway, this doesn't' harm.

Comment (helinwang) [inline go/master/service.go]: I think if a master gets restarted, it has no chance to fill in `JobTasks`, and later a pass finishes, todo queue will be set to `JobTasks` which is `nil`.

Comment (helinwang) [inline go/master/service.go]: You are right, just realized that we don't need `JobTasks` any more, replaced
```go
s.state.TODO = s.state.JobTasks
```
with:
```go
s.state.TODO = append(s.state.Done, s.state.Failed...)
```

File: go/master/service.go (hunk 2 above)
Comment (helinwang) [inline go/master/service.go]: Thanks! Done.

**Pull request merged**